Fix policy 000 warning

SebastianBoe opened the pull request:
Prepend the text 'cmake_minimum_required(VERSION 3.8.2)' into the
application and test build scripts.

Modern versions of CMake will spam users with a deprecation warning
when the toplevel CMakeLists.txt does not specify a CMake
version. This is documented in bug #8355.

To resolve this we include a dummy cmake_minimum_required() line into
the toplevel build scripts. This version is then overridden by
boileplate.cmake to avoid duplicating the true minimum required
version between all build scripts.

Also change CMake policy CMP000 from OLD to NEW which in turn
finally rids us of the verbose warning and fixes #8355.

The extra boilerplate is is considered more acceptable than the
verbosity of the CMP0000 policy.

jukkar commented on samples/application_development/external_lib/CMakeLists.txt:
```diff
@@ -1,3 +1,4 @@
+cmake_minimum_required(VERSION 2.4)
```
As all the samples seem to include boilerplate.cmake then why not add this cmake_minimum_required() there? This way this setting is always applied even for new samples and it is easy to set the required version in just one place.

SebastianBoe replied:
Because CMake does not allow this.

"a call in an included file is not sufficient"

https://cmake.org/cmake/help/v3.12/policy/CMP0000.html

jukkar replied:
Ok, perhaps that could be documented in the commit message. Now one starts to wonder why this change is needed unless one knows CMake well enough.

ulfalizer replied:
Would be nice to have it in the docs somewhere too. Looks cryptic otherwise.

SebastianBoe replied:
Added a clarifying CMake comment.

nashif commented:
so we are changing all of those samples and tests to only add a fake min required version that might end up confusing users more than the warning?

SebastianBoe commented:
> so we are changing all of those samples and tests to only add a fake min required version that might end up confusing users more than the warning?
-- @nashif 

Yes. I believe that one more line of boilerplate in every sample will cause less noise and confusion than the 8 or so lines that occur on every CMake invocation. Especially since CMP0000 presents itself as a warning.

ulfalizer commented:
IMO, seeing a spammy warning the first thing you do when you try out Zephyr gives a bad impression. I think people start to ignore output when there's too much of it too.

ulfalizer reviewed (changes requested):
http://docs.zephyrproject.org/application/application.html#creating-an-application will need to be updated as well, along with a motivation.

SebastianBoe committed:
cmake: change policy CMP0000 to NEW

Change CMake policy CMP0000 from OLD to NEW to resolve #8355.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
doc: Document new app boilerplate requirement

Document that the boilerplate code included in all applications must
now also include

cmake_minimum_required(VERSION 3.8.2)

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
> http://docs.zephyrproject.org/application/application.html#creating-an-application will need to be updated as well, along with a motivation.

Added the new line of boilerplate.

I don't think 'creating an application' is an appropriate place to document why cmake_minimum_require(2.4) is in the 'app' CMakeLists.txt.

These three lines:

```
      cmake_minimum_required(VERSION 2.4)
      include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
      project(NONE)

```

are boilerplate. Meaning, it is code that is intended to be copied mindlessly around without needing to understand what it does. It would only be necessary to understand their purpose if it were possible to opt-out of using it, but it is not, so the user does not need to make any informed decisions about it.

SebastianBoe commented:
> As a user, it bugs me when stuff like that is presented without motivation at least. Having people understand why stuff is there is nice.

The curious reader is free to remove the line. This will luckily not make a subtle change, but give an error with a clear error message. The not-so curious reader should not have to read about technical details outside of his control.

ulfalizer commented:
@SebastianBoe 
I like to know why particular steps are there, whether it's in IT or some other field. Nice to spread the knowledge around. Avoids cargo-culting later on too.

Could just be a short `.. note:` or the like.

SebastianBoe commented:
> I like to know why particular steps are there, whether it's in IT or some other field. Nice to spread the knowledge around. Avoids cargo-culting later on too.

Noted, but I believe the majority of users would prefer it as it is now.

ulfalizer commented:
What about a note like this?

```
The cmake_minimum_required line suppresses a warning from CMake about the top-level CMakeLists.txt not specifying a minimum version. The actual minimum CMake version required is specified in boilerplate.cmake.
```

It would not have been clear to me that that's what it's there for, even after removing the line and seeing the warning.

SebastianBoe commented:
Note that as of this PR it wouldn't be a warning, it would be a well-documented error,
so there is no possibility for user-error.

Updated accordingly with a note.

ulfalizer reviewed (approved):
LGTM

ulfalizer commented:
Guess it could be :file:\`CMakeLists.txt\` and :file:\`boilerplate.cmake\` as well, but not super important...

carlescufi commented:
@SebastianBoe can you fix CI?

carlescufi commented:
@SebastianBoe why not have the same minimum required in the top-level sample `CMakeLists.txt` as the one in the boilerplate file? i.e. why not 3.8.2 everywhere?

SebastianBoe commented:
> @SebastianBoe why not have the same minimum required in the top-level sample CMakeLists.txt as the one in the boilerplate file? i.e. why not 3.8.2 everywhere?

@carlescufi : Because then we would need to touch hundreds of samples every time we bump the version.

SebastianBoe commented:
> @SebastianBoe can you fix CI?

I don't think I can. CI is timing out because it is running many samples.

carlescufi commented:
>@carlescufi : Because then we would need to touch hundreds of samples every time we bump the version.

Makes sense. How about a comment in all `CMakeListst.txt` files like:

```
# This is not the minimum required version. See cmake/app/boilerplate.cmake for additional info.
cmake_minimum_required(VERSION 2.4)
```

carlescufi commented:
>I don't think I can. CI is timing out because it is running many samples.

Right, I have the exact same problem here: https://github.com/zephyrproject-rtos/zephyr/pull/9282. I wonder if @nashif can come up with a way to solve this problem.

SebastianBoe commented:
> Makes sense. How about a comment in all CMakeListst.txt files like:

I would rather omit this to minimize the amount of boilerplate.

The fact that 2.4 is not the minimum version is already documented in the application development guide. Also, mistakenly believing that 2.4 is the minimum version is not dangerous because

A) Following the guide will result in an appropriately installed version.
B) Not following the guide, and installing a CMake version that is too old will result in a clear error message.

pfalcon reviewed (approved):
I wonder if using "3.0" as a placeholder version would be at least a little bit less confusing (i.e. we require 3.x series).

Anyway, nobody ever thought that CMake is a nice, clean tool, so hail to workarounds.

dbkinder reviewed (approved):
doc change LGTM

ulfalizer commented:
CC @nashif

ulfalizer commented:
@SebastianBoe 
One issue might be that people could get the idea that they shouldn't use CMake features from after 2.4. Could freeze the boilerplate version at whatever the current minimum required CMake version is instead I guess.

Still an improvement as-is though.

nashif commented:
recheck

SebastianBoe committed:
cmake: Prepend 'cmake_minimum_required()' into 'app' build scripts

Prepend the text 'cmake_minimum_required(VERSION 3.8.2)' into the
application and test build scripts.

Modern versions of CMake will spam users with a deprecation warning
when the toplevel CMakeLists.txt does not specify a CMake
version. This is documented in bug #8355.

To resolve this we include a cmake_minimum_required() line into the
toplevel build scripts. Additionally, cmake_minimum_required is
invoked from within boilerplate.cmake. The highest version will be
enforced.

This patch allows us to afterwards change CMake policy CMP000 from OLD
to NEW which in turn finally rids us of the verbose warning.

The extra boilerplate is considered more acceptable than the verbosity
of the CMP0000 policy.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

pfalcon commented:
> To resolve the error this line of boilerplate must
be added to the top of the 'app' CMakeLists.txt file:

> cmake_minimum_required(VERSION 2.4)

Any replies to the suggestions sounded above to use "3.0" or the current real minimum version there (set once and frozen) ?

SebastianBoe commented:
> Any replies to the suggestions sounded above to use "3.0" or the current real minimum version there (set once and frozen) ?

Thank you for the feedback, but this has been addressed:

pfalcon commented:
> Thank you for the feedback, but this has been addressed before:
https://github.com/zephyrproject-rtos/zephyr/pull/9186#issuecomment-410650834

And the feedback provided takes that comment into account - there's no need "to touch hundreds of samples every time we bump the version", the talk is about selecting the least confusing version to set now, **once and for all**.

SebastianBoe commented:
I believe the CMake version will be bumped every couple of years.

I also believe 2.4 will be less confusing than 3.8, because then the version will never have reflected the true minimum version, as opposed to at some point be the minimum version, and then stopped being so.

pfalcon commented:
> I also believe 2.4 will be less confusing than 3.8

I agree, that's why I proposed 3.0 ;-). Anyway, I assume you really don't want to make any cosmetic changes to this PR, which can't be proven to improve anything, can understand that.

nashif commented:
I still do not understand why we have 2.4 is min required, why can't you have 

cmake_minimum_required(VERSION 3.8.2)

instead?
I just tried it, and it seems to work, or am I missing something here?

SebastianBoe commented:
> I still do not understand why we have 2.4 is min required, why can't you have
> 
> cmake_minimum_required(VERSION 3.8.2)
> 
> instead?
> I just tried it, and it seems to work, or am I missing something here?

Because 3.8.2 would imply that this line of boilerplate is denoting the minimum required version. If it does, then we will need to bump it when we bump the version, which would be a breaking change to user applications and require a modification to all samples.

Alternatively, we could not bump it when we bump the CMake version, which would be confusing to the user as the version will have two different meanings, first it meant the minimum version, then later on it did not.

I believe it is cleanest if this line is always just a static line of boilerplate to satisfy CMake's policy.

ulfalizer commented:
> Because 3.8.2 would imply that this line of boilerplate is denoting the minimum required version. If it does, then we will need to bump it when we bump the version, which would be a breaking change to user applications and require a modification to all samples.
> 
> Alternatively, we could not bump it when we bump the CMake version, which would be confusing to the user as the version will have two different meanings, first it meant the minimum version, then later on it did not.
> 
> I believe it is cleanest if this line is always just a static line of boilerplate to satisfy CMake's policy.

Guess it would only be potentially confusing for a short time at least.

nashif commented:
> Because 3.8.2 would imply that this line of boilerplate is denoting the minimum required version. 
But that is actually the case, putting something else is going to be wrong.


> If it does, then we will need to bump it when we bump the version, which would be a breaking change to user applications and require a modification to all samples.

I am not following, say we move to min required version of 3.12, what difference does it make if we had 2.4 or 3.8.2? At least we are going to be correct in terms of compatibility until we bump, with 2.4 we will be wrong immediately.

pfalcon commented:
> with 2.4 we will be wrong immediately.

We won't be wrong, the whole idea of `cmake_minimum_required(VERSION 2.4)` is to workaround cmake's ... well, strangeness. So, it doesn't matter what's written there, it's dummy. @SebastianBoe : I told that using "3.0" will be less confusing and easier to explain. Maybe ;-).

nashif commented:
> We won't be wrong, the whole idea of cmake_minimum_required(VERSION 2.4) is to workaround cmake's ... well, strangeness. So, it doesn't matter what's written there, it's dummy.

it is not visible as a workaround, there is no documentation or comments whatsoever saying it is a workaround, people here in this PR might know it is a workaround but someone opening the cmakefile of a sample or someone copying a sample and starting app will have 2.4 set as a the min required version which might lead to confusion. 
Still do not understand what we cant have 3.8.2 in the workaround which matches the true min. required version. Why did you pick 2.4?

SebastianBoe commented:
I see there is a misunderstanding / missing piece of the puzzle here:
There is a second invocation of cmake_minimum_required in boilerplate.cmake.

> But that is actually the case, putting something else is going to be wrong.

No, the boilerplate in the apps are not denoting the minimum version. Instead we have a second invocation of cmake_minimum_required in boilerplate.cmake that sets the version to 3.8.2. Since 3.8.2 is higher than 2.4 the enforced minimum version becomes 3.8.2.

> I am not following, say we move to min required version of 3.12, what difference does it make if we had 2.4 or 3.8.2?

The difference is that when we move to 3.12 all we need to do is bump the one invocation in boilerplate.cmake, instead the of every invocation.

>  there is no documentation or comments

Granted there is no comments where cmake_minimum_required(2.4) is invoked. But this is documented in the application guide[0].

This is documented in the application guide instead of as a comment near the invocation because  we would need to copy-paste that comment across all apps.

There is a lot of pushback on setting it to 2.4. I still believe the arguments for setting it at 2.4 hold, but I am OK with going against my judgement and setting it to the minimum version. But if we do so we should agree on what we do when the true version is bumped. E.g. bump it across all apps again, or leave it as a no-longer-accurate minimum version.

[0] https://github.com/zephyrproject-rtos/zephyr/pull/9186/files#diff-1a941310744bb7c02ac48a0a337ff4fbR115

nashif commented:
> There is a second invocation of cmake_minimum_required in boilerplate.cmake.

I know about this one, someone looking at zephyr from an application will only see the 2.4 though

> But if we do so we should agree on what we do when the true version is bumped. E.g. bump it across all apps again, or leave it as a no-longer-accurate minimum version.

Whatever we do would apply to both when we have 2.4 or 2.8.2, I am not sure why this would be different if we had 2.4 instead of 2.8.2. 

The main question, is there any technical reason why 2.8.2 would be wrong beside the argument above that we would need to change all files when we require a higher version?

ulfalizer commented:
From a practical perspective, I suspect a lot of people browsing the code won't find the explanation for what that `cmake_minimum_required()` boilerplate is doing there, and will assume that they can't use features more recent than CMake 2.4.

Think that might've bitten me.

SebastianBoe commented:
Changed the "app/CMakeList.txt" cmake version from 2.4 to 3.8.2 due to a general agreement that it should be set to 3.8.2.
